src/coreclr/gc/gc.cpp

```diff
@@ -22241,7 +22275,15 @@ end_no_gc_region_status gc_heap::end_no_gc_region()
         status = end_no_gc_alloc_exceeded;
 
     if (settings.pause_mode == pause_no_gc)
+    {
         restore_data_for_no_gc();
+        if (current_no_gc_region_info.soh_withheld_budget)
+        {
+            dprintf (1, ("[no_gc_callback] detaching callback on exit"));
+            current_no_gc_region_info.finalizer_action_kind |= finalizer_action_kind::detach_no_gc_callback;
```

Comment:
Should we unregister the callback unconditionally and immediately?

The API review notes https://github.com/dotnet/runtime/issues/66039#issuecomment-1105631442 say: "The callback is automatically removed when the no-GC region ends."

I would expect the following test to pass:
```
using System;

for (int i = 0; i < 100; i++)
{
    Console.WriteLine(i.ToString());
    if (!GC.TryStartNoGCRegion(100_000_000))
    {
       Console.WriteLine("TryStartNoGCRegion failed.");
       break;
    }
    GC.RegisterNoGCRegionCallback(50_000_000, () => Console.WriteLine("NoGCRegionCallback"));
    
    GC.EndNoGCRegion();
}
```
This test fails with InvalidOperationException today.